### PR TITLE
Some improvements

### PR DESCRIPTION
**Pull request by nkoshell**
## Changes 
- Examples try to get `API_KEY` from environment variables
- All dependencies moved to the `setup.py`
- Small refactor and reformat

## Additions
- API methods for `charge` resource (`cancel` and `resolve`)
- `APIAuthHeadersMixin` class for `APIAuth` class
- Internal method for building error class
- Alternative variant for registration API resource classes (class decorator)

### This changes are needed for [new async client](https://github.com/nkoshell/coinbase-commerce-python/tree/aiohttp-client/coinbase_commerce/aio)

### CODE REVIEW

File: coinbase_commerce/api_resources/charge.py [hunk 1]
```diff
@@ -1,8 +1,28 @@
...
+        return api_obj
+
+    @classmethod
+    def resolve(cls, entity_id, **params):
+        response = cls._api_client.post(
+            cls.RESOURCE_PATH, entity_id, 'resolve',
+            data=params
+        )
+        api_obj = util.convert_to_api_object(response, cls._api_client, cls)
+        return api_obj
```
Comment (chalk777) [inline coinbase_commerce/api_resources/charge.py]: return util.convert_to_api_object(response, cls._api_client, cls)

File: coinbase_commerce/api_resources/charge.py [hunk 2]
```diff
@@ -1,8 +1,28 @@
...
     RESOURCE_NAME = "charge"
+
+    @classmethod
+    def cancel(cls, entity_id, **params):
+        response = cls._api_client.post(
+            cls.RESOURCE_PATH, entity_id, 'cancel',
+            data=params
+        )
+        api_obj = util.convert_to_api_object(response, cls._api_client, cls)
+        return api_obj
```
Comment (chalk777) [inline coinbase_commerce/api_resources/charge.py]: return util.convert_to_api_object(response, cls._api_client, cls)

File: coinbase_commerce/api_resources/charge.py (hunk 1 above)
Comment (nkoshell) [reply coinbase_commerce/api_resources/charge.py]: Thanks

File: coinbase_commerce/api_resources/charge.py (hunk 2 above)
Comment (nkoshell) [reply coinbase_commerce/api_resources/charge.py]: Thanks

**Comment by guacamoli** [comment]
Thanks for the contribution, @nkoshell!

File: coinbase_commerce/util.py [hunk 3]
```diff
@@ -11,16 +11,36 @@
 RESOURCE_MAP = {}
 
 
+def register_resource_cls(*args, **kwargs):
+    def resource_decorator(cls):
+        rn = getattr(cls, "RESOURCE_NAME", resource_name)
```
[the remark this answers is not in the document]
Comment (itstehkman) [reply coinbase_commerce/util.py]: Nvm, I see it is defined below, but I think the getattr default should be defined as an explicit argument for clarity:

```
def resource_decorator(cls, resource_name_default=None):
```

[the remark this answers is not in the document]
Comment (nkoshell) [reply coinbase_commerce/util.py]: Ok, but in this case must explicit check keyword argument value 
```python
    if args and resource_name_default is not None:
        raise ValueError("cannot combine positional and keyword args")
    if len(args) == 1:
        return resource_decorator(args[0])
    elif len(args) != 0:
        raise ValueError("expected 1 argument, got %d", len(args))
    return resource_decorator
```
If in future we increase count keywords arguments, then code for this checking will be some bulky

File: examples/example_charge.py [hunk 4]
```diff
@@ -6,10 +6,11 @@
 metadata	          hash	Optional	Developer defined key value pairs
 redirect_url	    string	Optional	Redirect URL
 """
+import os
 
 from coinbase_commerce.client import Client
 
-API_KEY = "API_KEY"
+API_KEY = os.environ.get("COINBASE_COMMERCE_API_KEY", "API_KEY")
```
Comment (itstehkman) [inline examples/example_charge.py]: While I think this is helfpul, I think we should not do this incase somebody is using production keys in their environment variables - they accidentally try an example and make a destructive change that they did not want to do. We should let users playing around with the examples choose how to pick the API KEY for examples, and not default to envars.

File: coinbase_commerce/util.py (hunk 3 above)
[the remark this answers is not in the document]
Comment (itstehkman) [reply coinbase_commerce/util.py]: Hmm thats a good point, I'll leave it up to you. However, it seems like this decorator won't really change that much overtime, and wouldn't expect the # of keyword arguments to increase all that much.

File: coinbase_commerce/error.py [hunk 5]
```diff
@@ -99,23 +103,42 @@ class WebhookInvalidPayload(APIError):
     pass
 
 
-def build_api_error(response, blob=None):
+def _build_api_error(code, body=None, headers=None, blob=None,
+                     json_loads=json.loads):
```
Comment (itstehkman) [inline coinbase_commerce/error.py]: Is there ever any time somebody would want to specify the json_loads function? I don't think so, unless it is being mocked

File: examples/example_charge.py (hunk 4 above)
Comment (nkoshell) [reply examples/example_charge.py]: Ok, I revert this commit

Comment (itstehkman) [reply examples/example_charge.py]: Thank you!

**Pull request merged**